crates/uv-pypi-types/src/metadata/requires_dist.rs

```diff
@@ -86,3 +88,249 @@ impl RequiresDist {
+        if value.requires_dist.iter().all(|req| req.name != value.name) {
+            return Self {
+                name: value.name,
+                requires_dist: value.requires_dist,
+                provides_extras: value.provides_extras,
+            };
+        }
+
+        // Transitively process all extras that are recursively included.
+        let mut requires_dist = value.requires_dist.clone();
+        let mut seen = FxHashSet::<(ExtraName, MarkerTree)>::default();
+        let mut queue: VecDeque<_> = value
+            .requires_dist
+            .iter()
+            .filter(|req| req.name == value.name)
+            .flat_map(|req| req.extras.iter().cloned().map(|extra| (extra, req.marker)))
+            .collect();
+        while let Some((extra, marker)) = queue.pop_front() {
+            if !seen.insert((extra.clone(), marker)) {
+                continue;
+            }
+
+            // Find the requirements for the extra.
+            for requirement in &value.requires_dist {
+                if requirement.marker.top_level_extra_name().as_ref() == Some(&extra) {
+                    let requirement = {
+                        let mut marker = marker;
+                        marker.and(requirement.marker);
+                        Requirement {
+                            name: requirement.name.clone(),
+                            extras: requirement.extras.clone(),
+                            version_or_url: requirement.version_or_url.clone(),
+                            marker: marker.simplify_extras(std::slice::from_ref(&extra)),
+                            origin: requirement.origin.clone(),
+                        }
+                    };
+                    if requirement.name == value.name {
+                        // Add each transitively included extra.
+                        queue.extend(
+                            requirement
+                                .extras
+                                .iter()
+                                .cloned()
+                                .map(|extra| (extra, requirement.marker)),
+                        );
+                    } else {
+                        // Add the requirements for that extra.
+                        requires_dist.push(requirement);
+                    }
+                }
+            }
+        }
+
+        // Drop all the self-requirements now that we flattened them out, but retain any that have
+        // a version specifier.
+        requires_dist.retain(|req| {
+            if req.name == value.name {
+                if req
+                    .version_or_url
+                    .as_ref()
+                    .is_none_or(|version_or_url| match version_or_url {
+                        VersionOrUrl::Url(_) => false,
+                        VersionOrUrl::VersionSpecifier(specifiers) => specifiers.is_empty(),
+                    })
+                {
+                    return false;
+                }
```

Comment:
This is pretty weird, but we have a lot of lockfile tests (`lock_self_*`) that test cases in which a package depends on itself, and puts a version specifier on that dependency. I don't think we can drop those.

For example:

```toml
[project]
name = "project"
version = "0.1.0"
requires-python = ">=3.12"
dependencies = ["typing-extensions", "project==0.1.0"]
```

Reply:
I'm not fully comfortable with this whole situation, self-dependencies seem to be poorly specified right now...

For example, Hatchling drops the `==0.2.0` in the `foo` extra here:

```toml
[project]
name = "project"
version = "0.1.0"
requires-python = ">=3.12"
dependencies = []

[project.optional-dependencies]
foo = ["project[bar]==0.2.0"]
bar = ["iniconfig"]

[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"
```

Reply:
I tried but couldn't reproduce this:

```toml
[project]
name = "hatch-drops-self-specifier"
version = "0.1.0"

[project.optional-dependencies]
foo = ["project[bar]==0.2.0"]
bar = ["iniconfig"]

[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"
```

```shell
rm -rf venv
uv venv
uv build
uv pip install --find-links dist hatch-drops-self-specifier --no-cache
cat .venv/lib/python3.*/site-packages/hatch_drops_self_specifier-0.1.0.dist-info/METADATA
```

```
Metadata-Version: 2.4
Name: hatch-drops-self-specifier
Version: 0.1.0
Provides-Extra: bar
Requires-Dist: iniconfig; extra == 'bar'
Provides-Extra: foo
Requires-Dist: project[bar]==0.2.0; extra == 'foo'
```